RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,413 @@
+two direct subclasses of `torch.Tensor`. Both will return `NotImplemented`, and
+therefore, the check will fail and `TypeError` will be raised.
+
+Since subclasses are checked before superclasses in `__torch_function__`, it is
+guaranteed that the subclass implementation will be called first. In this
+instance, since `type(self)` is a subclass of all types, the code will
+continue. Since `type(self)` is not `torch.Tensor`, a view into the original
+data is created and returned.
+
+This also works for all operators: `__add__`, `__getitem__` and so on since in
+Python these operators are just dunder methods of the corresponding class.
+
+### Checking for compatibility
+One can check for compatibility with supported classes in the following manner:
+
+```python
+class MyTensor:
+    @classmethod
+    def __torch_function__(cls, func, types, args, kwargs):
+        HANDLED_CLASSES = (MyTensor, Tensor, ...)
+        if not issubclass(t, HANDLED_CLASSES) for t in types:
+            return NotImplemented
+        # Do further processing here.
+```
+
+### Implementing a subset of the API
+One can directly follow the following procedure to implement a subset of the
+API by using a hashmap to your own implementations of a function:
+
+```python
+_TORCH_IMPLEMENTATIONS = {}
+
+def implements(torch_function):
+    def inner(f):
+        _TORCH_IMPLEMENTATIONS[torch_function] = f
+        return f
+    return inner
+
+@implements(torch.add)
+def my_add(self, other):
+    # Implementation here
+
+class MyTensor:
+    @classmethod
+    def __torch_function__(cls, func, types, args, kwargs):
+        compatible = ...
+        if not compatible:
+            return NotImplemented
+        
+        if func not in _TORCH_IMPLEMENTATIONS:
+            return NotImplemented
+
+        return _TORCH_IMPLEMENTATIONS[func](*args, **kwargs)
+```
+
+### The need for `super().__torch_function__`
+To access super, one would do the following:
+```python
+class SubTensor(torch.Tensor):
+    @classmethod
+    def __torch_function__(cls, func, types, args, kwargs):
+        # Pre-processing here
+        val = super().__torch_function__(
+            func,
+            types
+            args,
+            kwargs
+        )
+        # Post processing here
+```
+
+To make the need for `super()` to be available concrete, let's consider the
+following scenario:
+
+```python
+class SubTensor(torch.Tensor):
+    @classmethod
+    def __torch_function__(...):
+        # Pre-processing
+        ret = super().__torch_function__(
+            func,
+            types
+            args,
+            kwargs
+        )
+        # Post processing
+        return ret
+
+class SubSubTensor(SubTensor):
+    def __add__(self, other):
+        # Pre-processing
+        ret = super().__add__(other)
+        # Post-processing
+        return ret
+```
+
+In this instance, with the proposed changes, processing would follow the
```

Comment:
try not to use "would ..." or "with proposed changes" - it's not clear if this is what _will_ happen, or if you're discussing some discarded alternative. just phrase it like "it works like a, b, c"